File: 121. Best Time to Buy and Sell Stock/main.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int max_profit = 0;
        int buy = prices[0];

        for (const auto &price : prices) {
            if (buy > price) {
                buy = price;
                continue;
            }

            int profit = price - buy;
            if (max_profit < profit) {
                max_profit = profit;
            }
        }

        return max_profit;
    }
};

#include <vector>

using std::max;
using std::vector;

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int max_profit = 0;
        int buy = prices[0];

        for (const auto &price : prices) {
            if (buy > price) {
                buy = price;
                continue;
            }

            max_profit = max(max_profit, price - buy);
        }

        return max_profit;
    }
};

File: 121. Best Time to Buy and Sell Stock/main.rs
struct Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max_prof = 0;
        let mut buy = prices[0];

        for price in prices {
            if (buy > price) {
                buy = price;
                continue;
            }

            max_prof = max_prof.max(price - buy);
        }

        max_prof
    }
}

use std::cmp::{max, min};

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        prices
            .iter()
            .fold((0, i32::MAX), |(profit, buy), &price| {
                (max(profit, price - buy), min(buy, price))
            })
            .0
    }
}
